src/Microsoft.DncEng.SecretManager/SecretManifest.cs

```diff
@@ -0,0 +1,103 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Text;
+using YamlDotNet.Core;
+using YamlDotNet.Serialization;
+
+namespace Microsoft.DncEng.SecretManager
+{
+    public class SecretManifest
+    {
+        public static SecretManifest Read(string filePath)
+        {
+            using var reader = new StreamReader(filePath, Encoding.UTF8);
+            return Parse(reader);
+        }
+
+        public static SecretManifest Parse(TextReader reader)
+        {
+            IDeserializer deserializer = new DeserializerBuilder()
+                .Build();
+
+            var parser = new Parser(reader);
+            return new SecretManifest(deserializer.Deserialize<Format>(parser));
+        }
+
+        #pragma warning disable IDE1006
+        // ReSharper disable All
+        private class Format
+        {
+            public Guid subscription { get; set; }
+            public string name { get; set; }
+            public bool missingSecretsAllowed { get; set; }
+            public Dictionary<string, Key> keys { get; set; }
+            public Dictionary<string, Secret> secrets { get; set; }
+
+            public class Key
+            {
+                public string type { get; set; }
+                public int size { get; set; }
+            }
+
+            public class Secret
+            {
+                public string type { get; set; }
```

Comment:
Are we expecting to expand this a bit in the future, like something like "manualRotationInstructions" or other things? Or do we expect this to be the full set?

Reply:
It might need to be expanded a bit, but we shouldn't need much more than this. "Manual rotation instructions" should be defined by the "type"

Reply:
"manual rotation instructions" can just launch a relevant wiki entry somewhere or a doc in core-eng or something.

Reply:
I guess that means there will be LOTS of types?  "connection-string-to-kusto-based-on-aad-app"? "github-user-account-password".

What about secrets that are related (like the password and otp for a single user, or the storage account keys and sas tokens based on them) Are those "parameters"?

Reply:
for user accounts we will probably have multiple entries in key vault, but one entry in the manifest. The "github-user-account" type would probably be a "composite" type of some kind that spits out multiple values. I don't think the kusto connection string one needs to be that long, there is only one type of connection string that is actual secret and needs to be rotated. MSI auth is just "auth to this kusto instance"

Reply:
Ooo, composite secrets. Very nice.  That should solve a lot of the problems. :-)  There are probably still a few one-off secrets that making a whole type for might be a lot of overhead (like the "admin key to grafana"... there will only ever be one of those).

I guess I don't know what sort of cost there is to making a new "type". If it's relatively straightforward, then maybe that's fine that you'd create a special type for your single secret.

Reply:
once i'm done a new type will just be "make a new implementation of this abstract class". Potentially we want a type that is just "some text with instructions" but I don't really like that.

Reply:
Seems like _most_ of them are at the very least "give some instructions, launch this URL, do... whatever the instructions said, and then paste the result from there". :-)

Is the potential of automating a browser for some of these something that is being considered? Or is that crazy?  Apparently "playwright" is in our internal feeds now, so it could theoretically drive edge. :-)

Reply:
It is something I want to do, but it will be at the end of the list.